Sources/Stores/Store.swift

```diff
@@ -119,77 +95,101 @@ public extension Store {
                                                                                         _ completion: @escaping StoreCompletionClosure<T>)
     -> Cancelable
     where Resource.T == T {
+
         let cancelable = StoreCancelable()
 
-        // fetch a fresh value from the network if no hit
-        func fetch(resource: Resource) {
-            cancelable.networkCancelable = networkStack.fetch(resource: resource) { (inner: () throws -> Data) -> Void in
-                do {
-                    let data = try inner()
+        // 1st - Fetch data from the Persistence
+        self.getPersistedData(for: resource) { (data) in
 
-                    guard cancelable.isCancelled == false else { return completion(nil, .cancelled, false) }
+            // If we have data we don't need to go to the network
+            if let data = data {
 
-                    // parse the new value from the data
-                    let value = try resource.parser(data)
+                // parse the new value from the data
+                self.process(data, fromCache: true, resource: resource, cancelable: cancelable, completion)
 
+            } else {
+
+                // 2nd - Try to fetch Data from Network
+                cancelable.networkCancelable = self.getNetworkData(resource) { (data, error) in
+
+                    // Check if it's cancelled
                     guard cancelable.isCancelled == false else { return completion(nil, .cancelled, false) }
 
-                    // update persistence with new value
-                    self.persist(data, for: resource)
-
-                    completion(value, nil, false)
-                } catch let Network.Error.url(error as NSError) where error.domain == NSURLErrorDomain
-                                                                   && error.code == NSURLErrorCancelled {
-                    completion(nil, .cancelled, false)
-                } catch let error as Network.Error {
-                    completion(nil, .network(error), false)
-                } catch let error as Parse.Error {
-                    completion(nil, .parse(error), false)
-                } catch {
-                    completion(nil, .other(error), false)
+                    // If we got an error we need to check if we have the data on disk
+                    if let error = error {
+
+                        return completion(nil, error, false)
+
+                    } else if let data = data {
+
+                        // parse the new value from the data
+                        self.process(data, fromCache: false, resource: resource, cancelable: cancelable, completion)
+                    }
```

Comment:
I know it's supposed to **never™** happen, but I would prefer if we added an `else` clause where we trigger a `fatalError`, since if that happens (due to any bug), the `completion` will never be called, for instance.